ENH: Reflect changes from numpy namespace refactor part 3

mtsokol opened the pull request:
Hi @rgommers,
Here's a new PR with follow-up work related to https://github.com/numpy/numpy/pull/24376. Basically, I update all items that are being removed in Part 3 of "main namespace refactor" PRs (Also, I took opportunity to update `np.double` to `np.float64` in Python files, as preferred canonical name).

rgommers commented on scipy/signal/_signaltools.py:
```diff
@@ -1569,7 +1569,7 @@ def medfilt(volume, kernel_size=None):
     order = numels // 2
 
-    if volume.dtype in [np.bool_, np.cfloat, np.cdouble, np.clongdouble,
+    if volume.dtype in [np.bool_, np.complex128, np.clongdouble,
```
also correct as a 1-for-1 replacement. I had to do a double take here - `cfloat` is not `complex64` as one would expect, but also `complex128`. 

That said, I think this code had a bug - it seems clear that the author intended to write `np.complex64` here (which is `np.csingle` rather than `np.cfloat`). Currently we get:
```python
>>> import numpy as np
>>> from scipy import signal
>>> x = np.linspace(0, 1, dtype=np.complex64)
>>> signal.medfilt(x)
File ~/code/scipy/build-install/lib/python3.10/site-packages/scipy/ndimage/_filters.py:1411, in _rank_filter(input, rank, size, footprint, output, mode, cval, origin, operation, axes)
   1409 input = numpy.asarray(input)
   1410 if numpy.iscomplexobj(input):
-> 1411     raise TypeError('Complex type not supported')
   1412 axes = _ni_support._check_axes(axes, input.ndim)
   1413 num_axes = len(axes)

TypeError: Complex type not supported
```

And after adding `np.complex64` to the list we'd get:
```python
  File ~/code/scipy/build-install/lib/python3.10/site-packages/scipy/signal/_signaltools.py:1574 in medfilt
    raise ValueError(f"dtype={volume.dtype} is not supported by medfilt")
ValueError: dtype=complex64 is not supported by medfilt
```

I think it'd be a valid bug fix to add `complex64` here.

mtsokol replied:
That's clear now - I changed used dtypes to sized aliases (both in `medfilt`'s implementation and test).

rgommers reviewed (commented):
This looks great from a first scan, thanks @mtsokol.

Perhaps this is also a good opportunity to start a `sed` script and put that under `tools/` in `numpy`? A lot of these changes can be automated (and I assume you do have some kind of script or tool already), and it'd be great to have something to point other users to that they can simply run over their own code.

mtsokol commented on scipy/interpolate/_fitpack2.py:
```diff
@@ -1917,8 +1917,6 @@ def __init__(self, theta, phi, r, w=None, s=0., eps=1E-16):
         if not 0.0 < eps < 1.0:
             raise ValueError('eps should be between (0, 1)')
 
-        if np.issubclass_(w, float):
-            w = ones(len(theta)) * w
```
@rgommers I claim that this `if` branch should be removed. 

`w` parameter is defined as `array_like, optional`. If it isn't None, then it's being set to `w = np.asarray(w)`. As a result `w` can end up as `None` or a numpy arrray.

Therefore, `np.issubclass_(w, float)` will always yield `False`, as only classes are accepted as arguments.  

Even hypothetically, if execution would have entered inside if statement, then:
1. `w` would have to be a class (`type(w) is type`)
2. Operation inside if statement would always fail: `np.ones(3) * float` raises `TypeError`.

rgommers replied:
That looks right, good catch.

mtsokol commented on scipy/interpolate/_fitpack2.py:
```diff
@@ -2070,8 +2068,6 @@ def __init__(self, theta, phi, r, tt, tp, w=None, eps=1E-16):
         if not 0.0 < eps < 1.0:
             raise ValueError('eps should be between (0, 1)')
 
-        if np.issubclass_(w, float):
-            w = ones(len(theta)) * w
```
Same comment as above.

mdhaber commented on scipy/stats/_mstats_extras.py:
```diff
@@ -15,7 +15,7 @@
 
 import numpy as np
-from numpy import float_, int_, ndarray
+from numpy import double, int_, ndarray
```
Why is it changing to `double` in this file?

mtsokol replied:
Because `float_` is being removed as it's an [other alias](https://numpy.org/devdocs/reference/arrays.scalars.html#other-aliases) - `double` is a dtype that `float_` aliases. I can also change it sized dtype, `float64`.

rgommers replied:
+1 for changing to `float64` here.

mtsokol replied:
Sure! I changed it to `float64`.

rgommers commented:
There's a few complaints from the linter, could you fix those @mtsokol? You can reproduce them locally with `python dev.py lint`.

mtsokol commented:
> There's a few complaints from the linter, could you fix those @mtsokol? You can reproduce them locally with `python dev.py lint`.

Linting errors came from the fact that I modified files which already had linting issues, therefore I skipped unrelated code. Sure! I fixed them.

rgommers reviewed (approved):
LGTM now, thanks @mtsokol!
